### PR TITLE
augmentation: add BandRotation for surface-EMG wristband layouts

### PR DESCRIPTION
**Pull request by bruAristimunha**
## Summary
- Adds `braindecode.augmentation.BandRotation` (and the underlying `band_rotation` function) — per-band circular roll along the channel axis plus inter-band temporal jitter, for surface-EMG inputs laid out as `(B, num_bands * electrodes_per_band, T)`.
- Models small wristband rotation between sessions and relative timing noise between two arms, as introduced in the emg2qwerty paper (Sivakumar et al., NeurIPS 2024).
- Slots into the existing `Transform` / `AugmentedDataLoader` pipeline; alphabetised in the `__init__` exports and changelog updated.

## Why this lives in braindecode
The augmentation is model-agnostic at the tensor level: any model that consumes a `(B, num_bands * electrodes_per_band, T)` EMG layout can use it. Keeping it in `braindecode.augmentation` lets non-Lightning users (Skorch / plain DataLoader) pick it up without dragging in a callback framework.

## API
```python
from braindecode.augmentation import BandRotation

transform = BandRotation(
    probability=0.5,
    num_bands=2,
    electrodes_per_band=16,
    band_offsets=(-1, 0, 1),
    max_temporal_jitter=120,  # 60 ms @ 2 kHz; paper value
    random_state=42,
)
```

## Performance note
A vectorized `torch.gather` collapsing the per-band loop into a single kernel was benchmarked and is ~16% **slower** for the typical `num_bands=2` case on CPU (the gather index tensor exceeds what two contiguous `torch.roll` calls touch); only past `num_bands >= 8` does the gather win. The loop is kept with a comment recording the benchmark so a future contributor doesn't redo the experiment.

## Test plan
- [x] 4 new unit tests in `test/unit_tests/augmentation/test_functional.py` (shape + seed reproducibility, no-op path, channel-count validation, circular-roll value preservation).
- [x] `BandRotation` added to the `test_set_params` parametrize list in `test_transforms.py` to exercise the Skorch `set_params` round-trip.
- [x] Full augmentation suite: `116 passed` (was 111 before; +5 from this PR).
- [x] All pre-commit hooks pass (ruff, mypy, isort, codespell, sphinx-lint).

## Changelog
Entry added under `Enhancements` in `docs/whats_new.rst`.

### CODE REVIEW

File: braindecode/augmentation/transforms.py [hunk 1]
```diff
@@ -1356,3 +1357,67 @@ def __init__(self, probability, interval=(0.5, 2), random_state=None):
...
+        Electrodes per band (e.g. ``16``).  Defaults to 16.
+    band_offsets : tuple of int, optional
+        Per-band roll values to sample from uniformly.  ``(-1, 0, 1)``
+        covers ±1-electrode misalignment.  Defaults to ``(-1, 0, 1)``.
+    max_temporal_jitter : int, optional
+        Max ±-sample temporal shift applied to band 1.  Defaults to 0
+        (jitter disabled).  The emg2qwerty paper uses 120 samples
+        (60 ms at 2 kHz).
+    random_state : int | numpy.random.Generator, optional
+        Seed for the rotation / jitter sampler.  Defaults to None.
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/augmentation/transforms.py]: Fixed in 12236961: docstring updated to ``int | numpy.random.RandomState | None`` to match ``sklearn.utils.check_random_state``'s actual return type.

File: braindecode/augmentation/transforms.py [hunk 2]
```diff
@@ -1356,3 +1357,67 @@ def __init__(self, probability, interval=(0.5, 2), random_state=None):
...
+        num_bands=2,
+        electrodes_per_band=16,
+        band_offsets=(-1, 0, 1),
+        max_temporal_jitter=0,
+        random_state=None,
+    ):
+        super().__init__(probability=probability, random_state=random_state)
+        self.num_bands = num_bands
+        self.electrodes_per_band = electrodes_per_band
+        self.band_offsets = tuple(band_offsets)
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/augmentation/transforms.py]: Fixed in 12236961: added explicit ValueError checks in ``BandRotation.__init__`` for ``num_bands >= 1``, ``electrodes_per_band >= 1``, non-empty ``band_offsets``, and ``max_temporal_jitter >= 0``.

File: braindecode/augmentation/functional.py [hunk 3]
```diff
@@ -1298,3 +1298,92 @@ def amplitude_scale(
...
+    expected_channels = num_bands * electrodes_per_band
+    if X.shape[1] != expected_channels:
+        raise ValueError(
+            f"X.shape[1]={X.shape[1]} != num_bands * electrodes_per_band="
+            f"{expected_channels}"
+        )
+
+    rng = check_random_state(random_state)
+    band_offsets_arr = np.asarray(band_offsets)
+    out = X.clone()
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/augmentation/functional.py:1394]: Fixed in 12236961: ``band_rotation`` now validates ``num_bands >= 1``, ``electrodes_per_band >= 1``, non-empty ``band_offsets``, that all offsets are integers, and ``max_temporal_jitter >= 0``. Each raises a labelled ValueError. Also addresses the wrap-around concern by adding ``circular_jitter: bool = True`` so callers can opt into crop-and-pad shifting.

File: test/unit_tests/augmentation/test_functional.py [hunk 4]
```diff
@@ -110,6 +111,62 @@ def test_amplitude_scale():
...
+        random_state=42,
+    )
+    out2, _ = band_rotation(
+        X, y, num_bands=2, electrodes_per_band=8,
+        band_offsets=(-1, 0, 1), max_temporal_jitter=4,
+        random_state=42,
+    )
+    assert out1.shape == X.shape
+    assert torch.equal(out1, out2)  # same seed → identical
+    assert not torch.equal(out1, X)  # something rolled
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply test/unit_tests/augmentation/test_functional.py]: Fixed in 12236961: test now forces ``band_offsets=(2,)`` so the rotation is non-zero by construction, decoupling the assertion from the RNG draw order.

File: braindecode/augmentation/transforms.py [hunk 5]
```diff
@@ -1356,3 +1357,95 @@ def __init__(self, probability, interval=(0.5, 2), random_state=None):
...
+        if num_bands < 1:
+            raise ValueError(f"num_bands must be >= 1, got {num_bands}")
+        if electrodes_per_band < 1:
+            raise ValueError(
+                f"electrodes_per_band must be >= 1, got {electrodes_per_band}"
+            )
+        band_offsets = tuple(band_offsets)
+        if not band_offsets:
+            raise ValueError("band_offsets must be non-empty")
+        if max_temporal_jitter < 0:
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/augmentation/transforms.py:1437]: Fixed in 4a87d943: `BandRotation.__init__` now also validates that `band_offsets` contains integers (`int` / `numpy.integer`), matching the existing check in `band_rotation`, so misconfigurations raise at Transform construction time.

File: test/unit_tests/augmentation/test_functional.py [hunk 6]
```diff
@@ -110,6 +111,142 @@ def test_amplitude_scale():
...
+    # |shift|-wide boundary region of band 1.
+    diff = (out_circ != out_pad).any(dim=1).squeeze(0)  # (T,)
+    n_diff = int(diff.sum())
+    assert 0 < n_diff <= common["max_temporal_jitter"]
+    # Padded variant has zeros somewhere on band 1's edge; circular has
+    # the original ramp values wrapped around.
+    band1_pad = out_pad[0, 2:, :]
+    band1_circ = out_circ[0, 2:, :]
+    assert (band1_pad == 0).any()
+    assert not (band1_circ == 0).any()
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply test/unit_tests/augmentation/test_functional.py]: Fixed in 4a87d943: the test is now conditional on the observed `n_diff`. When the RNG draws `shift == 0`, both modes are no-ops, so we assert `out_circ == out_pad == X` and return; otherwise we keep the boundary-region invariants. This decouples the assertions from the sampler's exact draw sequence.

File: test/unit_tests/augmentation/test_functional.py [hunk 7]
```diff
@@ -110,6 +111,142 @@ def test_amplitude_scale():
...
+
+
+def test_band_rotation_circular_roll_preserves_values():
+    """Per-band roll is circular: every input value should appear in the
+    output (no zero-padding) when offsets are non-trivial."""
+    # Small, deterministic input so the sorted-list comparison stays
+    # ``O(n log n)`` on a few-hundred-element tensor.  Don't scale this up
+    # to large shapes — for production-size inputs use a ``set``-based
+    # equality check instead.
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply test/unit_tests/augmentation/test_functional.py]: Fixed in 4a87d943: replaced the misleading "use a `set`-based equality check" hint with a pointer to a multiset-aware alternative (`torch.unique(..., return_counts=True)` / histogram comparison), and noted explicitly that a `set` would silently drop duplicates and miss value-multiplicity bugs.

File: braindecode/augmentation/functional.py [hunk 8]
```diff
@@ -1298,3 +1298,126 @@ def amplitude_scale(
...
+    .. [Sivakumar2024] Sivakumar, V., Seely, J., Du, A., Bittner, S. R.,
+       Berenzweig, A., Bolarinwa, A., Gramfort, A., & Mandel, M. I. (2024).
+       "emg2qwerty: A Large Dataset with Baselines for Touch Typing using
+       Surface Electromyography." *NeurIPS Datasets and Benchmarks Track*.
+    """
+    if num_bands < 1:
+        raise ValueError(f"num_bands must be >= 1, got {num_bands}")
+    if electrodes_per_band < 1:
+        raise ValueError(f"electrodes_per_band must be >= 1, got {electrodes_per_band}")
+    if not band_offsets:
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/augmentation/functional.py:1379]: Fixed in 92319afd: `band_rotation` now does `band_offsets = tuple(band_offsets)` before the non-empty / integer-element checks, so passing a `numpy.ndarray` (or any sequence-like) raises the labelled `ValueError` instead of numpy's ambiguous-truth-value error. Verified manually with empty/non-empty/float-array inputs.

File: braindecode/augmentation/transforms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/augmentation/transforms.py:1437]: Already addressed in 4a87d943 (just before this Copilot pass): `BandRotation.__init__` validates that `band_offsets` elements are integers right after `tuple(band_offsets)` and the non-empty check, mirroring `band_rotation`. This thread can be resolved.
